src/Storages/S3Queue/S3QueueOrderedFileMetadata.cpp

```diff
@@ -0,0 +1,257 @@
+    , buckets_num(buckets_num_)
+{
+}
+
+OrderedFileMetadata::Bucket OrderedFileMetadata::getBucketForPath(const std::string & path_, size_t buckets_num)
+{
+    return getBucketForPathImpl(path_, buckets_num);
+}
+
+static std::string getProcessorInfo(const std::string & processor_id)
+{
+    Poco::JSON::Object json;
+    json.set("hostname", DNSResolver::instance().getHostName());
+    json.set("processor_id", processor_id);
+
+    std::ostringstream oss;     // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+    oss.exceptions(std::ios::failbit);
+    Poco::JSON::Stringifier::stringify(json, oss);
+    return oss.str();
+}
+
+OrderedFileMetadata::BucketHolderPtr OrderedFileMetadata::tryAcquireBucket(
+    const std::filesystem::path & zk_path,
+    const Bucket & bucket,
+    const Processor & processor)
+{
+    const auto zk_client = getZooKeeper();
+    const auto bucket_lock_path = zk_path / "buckets" / toString(bucket) / "lock";
+    const auto processor_info = getProcessorInfo(processor);
+
+    zk_client->createAncestors(bucket_lock_path);
+
+    auto code = zk_client->tryCreate(bucket_lock_path, processor_info, zkutil::CreateMode::Ephemeral);
+    if (code == Coordination::Error::ZOK)
+        return std::make_shared<BucketHolder>(bucket_lock_path, zk_client);
+
+    if (code == Coordination::Error::ZNODEEXISTS)
+        return nullptr;
+
+    if (Coordination::isHardwareError(code))
+        return nullptr;
+
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected error: {}", magic_enum::enum_name(code));
+}
+
+std::pair<bool, IFileMetadata::FileStatus::State> OrderedFileMetadata::setProcessingImpl()
+{
+    /// In one zookeeper transaction do the following:
+    enum RequestType
+    {
+        /// node_name is not within failed persistent nodes
+        FAILED_PATH_DOESNT_EXIST = 0,
+        /// node_name ephemeral processing node was successfully created
+        CREATED_PROCESSING_PATH = 2,
+        /// max_processed_node version did not change
+        CHECKED_MAX_PROCESSED_PATH = 3,
+    };
+
+    processing_id = node_metadata.processing_id = getRandomASCIIString(10);
+    const auto zk_client = getZooKeeper();
+    while (true)
+    {
+        NodeMetadata processed_node;
+        Coordination::Stat processed_node_stat;
+        bool has_processed_node = getMaxProcessedFile(processed_node, &processed_node_stat, zk_client);
+        if (has_processed_node)
+        {
+            LOG_TEST(log, "Current max processed file {} from path: {}",
+                        processed_node.file_path, processed_node_path);
+
+            if (!processed_node.file_path.empty() && path <= processed_node.file_path)
+            {
+                return {false, FileStatus::State::Processed};
+            }
+        }
+
+        Coordination::Requests requests;
+        requests.push_back(zkutil::makeCreateRequest(failed_node_path, "", zkutil::CreateMode::Persistent));
+        requests.push_back(zkutil::makeRemoveRequest(failed_node_path, -1));
```

Comment:
I do not use it explicitly in any of the many cases it could be used in s3queue, because as there are many requests here and we need to process their responses (log or throw exception) and if we use `addCheckNotExistsRequest` - it could be either one request (if certain feature flag is enabled) or two requests, which makes it more obscure about the number of responses we get. Now we can see that there are two requests and we clearly see how we need to check for responses.

Btw I notice this file was mostly reviewed last week and this review is on top of the outdated version of the file - I pushed significant changes to it on friday after marking the pr as ready for review, so please verify that you checked them.